src/mirror.rs

```diff
@@ -0,0 +1,65 @@
+use url::{Url, ParseError};
+use std::collections::HashSet;
+
+pub struct MirrorContext {
+    base_urls: Vec<Url>,
```

Comment:
What's your stance on storing parsed URLs?  I figured earlier is probably better since it eliminates an allocation and moves the errors closer to where the dev made the mistake.